### ISSUE TITLE
Allow custom TextSelectionControls inside SelectableText

### ISSUE THREAD

**Issue by v7lin** [flutter/flutter#36985]
## Use case

Custom Menu for SelectableText

**Comment by jamesblasco** [comment]
![image](https://user-images.githubusercontent.com/19904063/84591740-1a906780-ae41-11ea-9e20-5795d589c173.png)

## Use case

Right now it is only possible to have the Cupertino and material appearance of the selection control. And this is limited to the Copy/Cut/Paste/Select All options. 

Having the ability to customize the selection control would bring us more freedom to build custom solutions.
Do you want to search on google a selected text? Or call a selected phone number? Highlight the selected text? Make it bold?

The implementation doesn't look difficult as it would be just adding the optional param and use it if it is not null and isSelectabe = true. [Check here were to add it](https://github.com/flutter/flutter/blob/97dc7eee570a31ed34f10834abf20549184cfe35/packages/flutter/lib/src/material/selectable_text.dart#L624).

**Comment by jerryan999** [comment]
waiting for this feature

**Comment by guidezpl** [comment]
Closing, as duplicate of #31412

**Comment by jamesblasco** [comment]
@guidezpl Besides they are similar, I don't think they are the same.

The issue  #31412 is to implement new options to the current materialTextSelectionControls/cupertinoTextSelectionControls.

This issue is about allowing developers to implement our own solution of TextSelectionControl besides the already implemented ones materialTextSelectionControls/cupertinoTextSelectionControls.

**Comment by guidezpl** [comment]
Ah sorry, I understand now. Seems like a reasonable ask. Basically plumbing SelectableText to https://api.flutter.dev/flutter/widgets/EditableText/selectionControls.html from Selectable. We may want to also want this in TextField and CupertinoTextField. @justinmc What do you think?

**Comment by justinmc** [comment]
Yeah it makes sense to me that the selection menu should be totally customizable.  If we do simply pipe through a selectionControls parameter, I'm not sure how nice the experience will be of creating a custom TextSelectionControls will be right now.  We'll have to try it out and see, maybe there is more we can do to make this approachable.

Plus I do plan to fix https://github.com/flutter/flutter/issues/31412 when I have time.

**Comment by jamesblasco** [comment]
I think piping selectionControls param would allow a high level of personalization but not very intuitive for most developers.  I would personally that option and in a close future create a TextSelectionControls that simplifies the process for the general case of a `DecoratedBox/CustomWidget` that allows a row of widgets inside that can interact with the text. (eg. Medium controls).

![image](https://user-images.githubusercontent.com/19904063/91598964-cdaf2e00-e965-11ea-8ccf-e7fcb3a9a868.png)

After I finish my GSoC project this week, I will try to create my own TextSelectionControls and provide some extra feedback.

**Comment by justinmc** [comment]
@jamesblasco Cool, let me know if you come up with anything!  I think the idea of a selectionControls param plus a TextSelectionControls that's more easier to work with could be nice.  Maybe we could provide the default-looking Android and iOS selection menu buttons and dividers as public widgets.

**Comment by jamesblasco** [comment]
<img align="right"  height="600" src="https://user-images.githubusercontent.com/19904063/94375942-7d1f2200-0117-11eb-8e87-e5ca6e07b613.gif"/> I made some progress with the idea I have in mind;

Here is the flutter version I am using with the changes needed: https://github.com/flutter/flutter/pull/66785
Here is the project with the TextSelectionControls: https://github.com/jamesblasco/selection_controls_example

The `main.dart` file contains an example from the code inside 'src' folder

This is a simple case with the `DefaultTextSelectionControls`:
```dart
 DefaultTextSelectionControls(
      handle: MaterialSelectionHandle(),
      toolbar: MaterialSelectionToolbar(
        backgroundColor: Colors.black,
        shape: BeveledRectangleBorder(
            borderRadius: BorderRadius.circular(12), side: BorderSide()),
        theme: ThemeData.dark(),
        items: [
          TextSelectionToolbarItem(
            title: Icon(Icons.brush, size: 18),
            onPressed: (selectionController) {
              final selection = selectionController.selection;
              // Do something when pressed
            },
          ),
          TextSelectionToolbarItem.copy(),
          TextSelectionToolbarItem.paste()
        ],
      ),
    );

```
The available handles are MaterialSelectionHandle are CupertinoSelectionHandle, but others can be created.
You can use the constructor `DefaultTextSelectionControls.withPlatformHandler` to use the handler that matches the device os.

All `CupertinoSelectionToolbar`, `MaterialSelectionToolbar` and your own implementation of `TextSelectionToolbar` uses `TextSelectionToolbarItem` to define the options of the toolbar. You can use a predefined action, your own action with the predefined UI or build it with your own new design (eg. Row with title and icon)

The idea is that they have the cut, copy, paste, selectAll options by default in that order.

<img align="left"  height="150" src="https://user-images.githubusercontent.com/19904063/94376157-16026d00-0119-11eb-8be1-1bcee5073523.png"/> 

While `CupertinoSelectionToolbar` and `MaterialSelectionToolbar` will hide the options that are not available(eg. paste when there is nothing to paste), it is also possible also to show them and implement a disabled UI.

 `MaterialSelectionToolbar` allows defining a custom background color, shape, clipBehaviour elevation and theme. (Inspired by the options used in showModalBottomSheet to customize the Material object inside.

@justinmc let me know what you think. If this is a bit out of scope to implement inside Flutter, at least with the PR this could be published as a package.

**Comment by justinmc** [comment]
@jamesblasco This is seriously awesome work.  I got your demo running locally and it looks and feels great.

That Github repo would make a great pub package as it is right now, but I think Flutter should indeed bring a lot/all of this functionality internal.  We have planned to support custom menu items for awhile, and with new desktop platforms to support, we may need to create additional menu styles as well.

I've been promising to write a design doc for this work for a long time.  If I go ahead and write that, can I tag you for some collaboration?  I'll write up all of the requirements we have for our text selection menus, and then we can figure out how to achieve them technically and how your project might fit the goals.  Let me know if you're up for that!

**Comment by jamesblasco** [comment]
Thanks for your kind words!. I would be up for the challenge! Let me know how I can help.

That was just a one-day work but I still think it covers the main ideas I had. I need to polish some parts and the next step will be to create a custom SelectionToolbar with an appearance different from the already built ones.  

I can't make TextSelectionControls work with a desktop platform. Am I missing something or is it not supposed to work yet?

**Comment by justinmc** [comment]
The text selection menus are disabled on desktop platforms right now, I believe.  The only way to copy is to use ctrl-c.  In the future we will either find a way to display the native menu, or create an identical menu in Flutter.

**Comment by justinmc** [comment]
@jamesblasco Here's the design doc: https://docs.google.com/document/d/1hyay1TyUZB7DXzWw45-ruGMZM1yCMWEnC3DzOJYemuc/edit#

I still need to elaborate on the implementation details, but could you look at the Overview section when you have time and tell me if you can think of any other use cases we need to support?

**Comment by SeasonLeee** [comment]
@justinmc hey, I just wonder how you get @jamesblasco 's example running?

I clone his version's flutter, but failed to set the sdk path to that in Android Studio with this error:
![image](https://user-images.githubusercontent.com/13610283/95722187-2df3f800-0ca6-11eb-9e46-1d1ceb83613a.png)

Any idea how to deal with this?

**Comment by justinmc** [comment]
@SeasonLeee I believe we talked a bit on Discord, but please keep me updated with what your company needs out of the text selection menu customization!

**Comment by jamesblasco** [comment]
Hey @justinmc, I have updated my example with some new ideas. `TextSelectionToolbars` have become '[ContextMenus'](https://github.com/jamesblasco/selection_controls_example/blob/context_menu/lib/context_menu.dart#L7) and they can now be used both inside and EditableText widget and in another widget ( similar to PopupMenuButton). If they are used as TextSelectionToolbars they will have access to the clipboard and text selection features.


Apart from the material and cupertino TextSelectionToolbars, I also started with 
 - `CascadeContextMenu` (ref: https://twitter.com/saketme/status/1314834865469222917?s=20) 
 - `CupertinoPullDownMenu` (ref: https://github.com/flutter/flutter/issues/60298)
I also started playing with nested menus and I would like to make the current `ContextMenuItems` become Widgets, but keeping the flexibility they have now.Still a lot of work to do but I think it is good for a general idea.
https://github.com/jamesblasco/selection_controls_example/tree/context_menu


![](https://im3.ezgif.com/tmp/ezgif-3-77fb4e192f30.gif)

**Comment by SeasonLeee** [comment]
@justinmc Yeah, sure thing! And thanks for all the detailed responses, you are the best😆

@jamesblasco and man, your work is phenomenal!!!! Thanks for all your contributions!!! I will have a look on that!

**Comment by justinmc** [comment]
@jamesblasco Awesome, thank you for the update!  I'm getting back to looking at this today.  We may go forward with your PR https://github.com/flutter/flutter/pull/66785 for now while I continue my work in investigating how to bring your solution into the framework.  That PR would unblock @SeasonLeee.  I'll respond soon when I have a better understanding of the changes.

**Comment by justinmc** [comment]
I think we should go forward with https://github.com/flutter/flutter/pull/66785.  I've left a comment on it.

**Comment by justinmc** [comment]
@jamesblasco I've merged https://github.com/flutter/flutter/pull/69428, which makes it much easier to build a custom text selection menu without needing to duplicate quite as much code. The API is still not ideal, but it's much better.

I have an example of how to build a custom menu here: https://github.com/justinmc/flutter-text-selection-menu-examples/blob/master/lib/custom_menu_page.dart

Next I'll try to do the same for iOS, so I'l keep this issue open until then.

**Comment by jamesblasco** [comment]
Thanks, @justinmc! Great work! I have been following your example repo for a while and I think the overall api is good.

Maybe it would be great to hide the logic from the clipboard and Paste/Copy/Cut/Select handlers and make it accessible from the context through an InheritedWidget.

**Comment by justinmc** [comment]
Do you mean the handles for the default text selection menu buttons, like this? https://github.com/flutter/flutter/blob/38fe7e2b1ca62433634bf6f3efce5f65de9914a7/packages/flutter/lib/src/widgets/text_selection.dart#L238

**Comment by jamesblasco** [comment]
Yess, maybe moving to a widget together with a `StatefulWidget` that listens to `ClipboardStatusNotifier` changes. 
As the TextSelectionDelegate will be the same for that widget, we could move these functions from that class to the widget tree.

So from your example:
https://github.com/justinmc/flutter-text-selection-menu-examples/blob/master/lib/custom_menu_page.dart

```dart

handlePaste: canPaste(delegate) && handlePaste != null ? () => handlePaste(delegate) : null,
----
if (widget.handlePaste != null && widget.clipboardStatus.value == ClipboardStatus.pasteable)
  _TextSelectionToolbarItemData(
    label: localizations.pasteButtonLabel,
    onPressed: widget.handlePaste,
)
```


Do like:

```dart
final controller = TextSelectionToolbarController.of(context);
if(controller.canPaste) 
  _TextSelectionToolbarItemData(
    label: localizations.pasteButtonLabel,
    onPressed: controller.handlePaste,
  )
```

In the example I made I have something similar to what I am saying here:
https://github.com/jamesblasco/selection_controls_example/blob/master/lib/src/selection_toolbar_controller.dart

**Comment by YeungKC** [comment]
We probably don't need TextSelectionControls, which is probably a different idea.

Because depending on the platform, the toolbar doesn't necessarily show up after it's selected, it could show the toolbar when you click on a blank space, or it could be a special design (like GitHub @ someone's autocomplete).

I was wondering if I could have EditableText notify AncestorWidget of the cursor size and position.
Then some AncestorWidget would receive this data and draw a different toolbar as needed (we'd need to implement the default MD style and Cupertion style, of course). And the AncestorWidget can also control the show and hide of the toolbar, and can distinguish between different trigger events.

I think this design will be more flexible.

cc @justinmc  @jamesblasco

**Comment by justinmc** [comment]
Let's continue this discussion in https://github.com/flutter/flutter/issues/73574.
